### ISSUE TITLE
Ensure that clients are able to get their results even if QPU-ID accessors change

### ISSUE THREAD

**Issue by BatmanAoD** [rigetti/qcs-sdk-rust#341]
We need a way to ensure that clients are able to get their results from a job, even if the QPU-ID used to submit the job gets a new accessor.

Currently, `retrieve_results` uses the QPU-ID to look up the address to which it sends the get-results request. But if QPU-IDs are not uniquely correlated with result queues, then this can return an address pointing to a service that doesn't actually know about the specific job in question. In other words, if a new accessor corresponding to a different ConServ instance is added to the network after an executable is submitted but before the client requests the result, and if the client isn't using a cached accessor, then the client will be told that the job doesn't exist.

### Planned solution

We will store some kind of unique accessor info in the job-handle itself, and require that for all result lookups.

### Other solutions considered

* Use an accessor-caching TTL that's long enough to outlive most jobs. This reduces the risk but of course doesn't eliminate it (for instance, the client can easily save the job-ID somewhere persistent and then bounce their client; or, a job with a very high number of shots could outlive the TTL).
* Create some kind of global-result-queue mechanism, where the results for a given QPU-ID are stored in the same queue regardless of the accessor (so, for instance, simulated jobs would go to the same queue).

(relates to #336)

**Comment by jselig-rigetti** [comment]
IMO it's up to the server side to provide this stability, but it does seem reasonable to cache the connection endpoint with the job handle.

**Comment by BatmanAoD** [comment]
Yeah... "if QPU-IDs are not uniquely correlated with result queues" sounds like a bizarrely pathological configuration. But I do think the proposed endpoint-caching makes sense.
